test/legacy_test/test_moe_permute_unpermute.py

```diff
@@ -0,0 +1,192 @@
+
+
+def fabricate_dispatch_result(
+    seqlen,
+    token_length,
+    topk,
+    num_experts,
+    data_type="bfloat16",
+    broadcast_ratio=0.5,
+):
+    """Helper function to generate test data."""
+    tokens = paddle.randn([seqlen, token_length], dtype=data_type)
+
+    tokens_scale = paddle.empty([0])
+    if data_type == "float8_e4m3fn":
+        scale_cols = (token_length + 127) // 128
+        tokens_scale = paddle.randn([seqlen, scale_cols], dtype="float32")
+
+    # Calculate expert counts with normal distribution
+    expected_experts = max(1, min(broadcast_ratio * num_experts, topk))
+    std_dev = max(1, expected_experts / 6)
+    experts_count = paddle.normal(expected_experts, std_dev, [seqlen])
+    experts_count = paddle.clip(
+        paddle.round(experts_count), 1, min(topk, num_experts)
+    )
+    experts_count = paddle.cast(experts_count, "int32")
+
+    # Preallocate results
+    dispatched_indices = paddle.full([seqlen, topk], -1, dtype="int32")
+    dispatched_probs = paddle.zeros([seqlen, topk], dtype="float32")
+
+    # Batch generate expert indices and probabilities
+    for i in range(seqlen):
+        count = experts_count[i].item()
+        indices = paddle.randperm(num_experts)[:count]
+        dispatched_indices[i, :count] = indices
+        prob_value = 1.0 / count
+        dispatched_probs[i, :count] = paddle.full(
+            [count], prob_value, dtype=data_type
+        )
+
+    # Calculate expert token counts
+    valid_indices = dispatched_indices.reshape([-1])
+    valid_mask = valid_indices >= 0
+    valid_experts = valid_indices[valid_mask]
+    expert_counts = paddle.histogram(
+        valid_experts, bins=num_experts, min=0, max=num_experts - 1
+    )
+    expert_counts = paddle.cast(expert_counts, "int32")
+    expert_counts = list(expert_counts)
+
+    return (
+        tokens,
+        tokens_scale,
+        dispatched_indices,
+        dispatched_probs,
+        expert_counts,
+    )
+
+
+def tensor_max_abs_rel_err(a, b, eps=1e-8):
+    """Calculate max absolute and relative error between two tensors."""
+    max_abs_err = paddle.max(paddle.abs(a - b))
+    denom = paddle.maximum(paddle.abs(a), paddle.abs(b))
+    denom = paddle.maximum(denom, paddle.to_tensor(eps, dtype=denom.dtype))
+    max_rel_err = paddle.max(paddle.abs(a - b) / denom)
+    return max_abs_err, max_rel_err
+
+
+class TestFusedMoePermuteUnpermute(unittest.TestCase):
+    """Test cases for moe_permute and moe_unpermute."""
+
+    SEQLEN = 16384
+    TOKEN_LEN = 7168
+    DTYPES = ["bfloat16"]
+    EXPERT_NUMS = [4, 8]
+    TOPKS = [4, 8]
+
+    def setUp(self):
+        """Initialize test environment."""
+        paddle.seed(42)  # For reproducibility
+
+    def test_permute_unpermute_consistency(self):
+        """Test that permute + unpermute recovers original tensors."""
+        for dt in self.DTYPES:
+            for expert_num in self.EXPERT_NUMS:
+                for topk in self.TOPKS:
```

Comment:
```suggestion
        for (dt, expert_num, topk) in itertools.product(self.DTYPES, self.EXPERT_NUMS, self.TOPKS):
```

使用 `itertools.product` 折叠多个 for 循环，减少锁进层级